File: SystemIdentification/SystemIdentification.py
import numpy as np
import csv
import os
import matplotlib.pyplot as plt
from scipy import signal


# Takes in output motor speed in counts/s
# motor_speed =  pos - pos_prev/delta_time
# G_P_dot(s) = (lambda/T)/(s_1/T)

# lambda = max(motor_speed)/max(input)
# T = the time at which this is true = 0.632 * steady_state_value

delta_time = 2000 # Micro Seconds
delta_time = 2000/1e6 # Conv to seconds



def store_recorded_data(path):
    pos1,pos2,in1,in2 = [],[],[],[]

    with open(path,'r') as file:
        reader = csv.reader(file)

        for row in reader:
            pos1.append(float(row[0]))  # Store values from the first column
            pos2.append(float(row[1]))  # Store values from the second column
            in1.append(float(row[2]))  # Store values from the third column
            in2.append(float(row[3]))  # Store values from the fourth column
    pos1 = np.array(pos1)
    pos2 = np.array(pos2)
    in1 = np.array(in1)
    in2 = np.array(in2)
    return pos1,pos2,in1,in2




def get_time_arr(arr):
    arrLen = np.size(arr)
    time = np.linspace(0,arrLen * delta_time,arrLen)
    return time

def getMotorSpeedArr(pos):
    motor_speed_arr = np.array([])
    for i,x in enumerate(pos):
        if i == 0:
            motor_speed_arr = np.append(motor_speed_arr,0)
        else:
            
            motor_speed_arr = np.append(motor_speed_arr,(pos[i] - pos[i-1])/delta_time)

    return motor_speed_arr

def getMaxMotorSpeed(motor_speed_arr): 
    return np.max(motor_speed_arr)
    

def getLambda(maxMotorSpeed, maxInput):
    return maxMotorSpeed/maxInput



def getTimeConstantT(steadyStateValue, motor_speed_arr, time_arr):
    threshold = 0.632 * steadyStateValue
    T_index = np.where(motor_speed_arr >= threshold)[0][0]
    T = time_arr[T_index]
    return T


def getIdealResponse(Lambda,T,inputArr,delta_time=delta_time):
    numerator = [Lambda/T]
    denominator = [1,1/T]
    system = signal.TransferFunction(numerator,denominator,dt=delta_time)
    tout, output = signal.dlsim(system,inputArr)
    return output

def getControlConsants(T,Lambda,Ts): # Ts Desired closed-loop settling time, T settling time
    w_cl = 7.55/Ts
    kp = (3*w_cl**2 * T)/Lambda
    ki = (w_cl**3 * T)/Lambda
    kd = (3 * w_cl * T-1)/(Lambda)

    return kp, ki, kd
file_name = "recorded_data.csv"
dirname = os.path.dirname(__file__)
path = os.path.join(dirname,file_name)


pos1,pos2,in1,in2 = store_recorded_data(path)
time = get_time_arr(pos1)
# An error is recording before 1000 means I have to do this
time = time[1000:]
pos1 = pos1[1000:]
in1 = in1[1000:]
motor_speed_arr1 = getMotorSpeedArr(pos1)
max_motor_speed1 = getMaxMotorSpeed(motor_speed_arr1)
maxInput1 = getMaxMotorSpeed(in1)
Lambda1 = getLambda(max_motor_speed1,maxInput1)
# Lambda = 22.5
T1 = getTimeConstantT(max_motor_speed1,motor_speed_arr1,time)
print(T1)
idealMotorSpeed1 = getIdealResponse(Lambda1,T1,in1)
# T = 0.29
kp1,ki1,kd1 = getControlConsants(T1,Lambda1,Ts=0.35)
print(f'{kp1 = }, {ki1 = }, {kd1 = }')


pos2 = pos2[1000:]
in2 = in2[1000:]
motor_speed_arr2 = getMotorSpeedArr(pos2)
max_motor_speed2 = getMaxMotorSpeed(motor_speed_arr2)
maxInput2 = getMaxMotorSpeed(in2)
Lambda2 = getLambda(max_motor_speed2,maxInput2)
# Lambda = 22.5
T2 = getTimeConstantT(max_motor_speed2,motor_speed_arr2,time)
idealMotorSpeed2 = getIdealResponse(Lambda2,T2,in2)

# T = 0.29
kp2,ki2,kd2 = getControlConsants(T2,Lambda2,Ts=0.35)
print(f'{kp2 = }, {ki2 = }, {kd2 = }')

# plt.plot(time,motor_speed_arr1)
# plt.show()
fig,(ax1,ax2,ax3) = plt.subplots(3,1)
ax1.plot(time,motor_speed_arr1, '-',label = "recorded" )
ax1.plot(time,idealMotorSpeed1, '-',label = 'ideal')
ax1.legend()

ax2.plot(time,motor_speed_arr2)
ax2.plot(time,idealMotorSpeed2)

ax3.plot(time,in1)



plt.show()

File: SystemIdentification/systemIdentification.m
filename1 = "systemIdentification.csv";
filename2 = "systemIdentification.csv";

startIdx = 700;
A = readmatrix(filename1);
B = readmatrix(filename2);

dt = 2000/1e6;
pos1 = A(startIdx:end,2);
pos2 = B(startIdx:end,2);
in1 = A(startIdx:end,1);
in2 = B(startIdx:end,4);

smoothed_position = sgolayfilt(pos1, 3, 15); % Polynomial order 3, window size 15
speed1 = [0;diff(smoothed_position)/dt];

speed2 = [0;diff(pos2)/dt];
time = linspace(0, (length(pos1)) * dt, length(pos1));

data1 = iddata(speed1,in1,dt);
sys1 = tfest(data1,1);

data2 = iddata(speed2,in2,dt);
sys2 = tfest(data2,1);
%G2 = overS * sys2 / (Gc * (1 - sys2 )); % Finding plant system from closed loop response

time = 0:dt:2613*dt;

figure
plot(time,in1,'Linewidth',1.5)
xlabel('Time (s)')
ylabel('Input u (PWM)')

figure
compare(data1,sys1);
ylabel('Motor speed (counts/s)')
xlabel('Time')

T1 = 1/sys1.Denominator(2)
Lambda1 = sys1.Numerator * T1

%T2 = 1/sys2.Denominator(2)
%Lambda2 = sys2.Numerator * T1

Ts = 0.35;
u_max = max(in1)
[kp1,ki1,kd1] = getControlConstants(T1, Lambda1, Ts)


%[kp2,ki2,kd2] = getControlConstants(T2, Lambda2, Ts)
Gc = tf([kd1,kp1,ki1],[1,0,0])
G = sys1 * Gc;

% margin(G)

function [kp, ki, kd] = getControlConstants(T, Lambda, Ts)
    % getControlConstants calculates control constants for a control system.
    % 
    % Inputs:
    %   T      - Settling time (plant)
    %   Lambda - Control tuning parameter
    %   Ts     - Desired closed-loop settling time
    %
    % Outputs:
    %   kp - Proportional gain
    %   ki - Integral gain
    %   kd - Derivative gain

    % Calculate closed-loop natural frequency
    w_cl = 7.55 / Ts;

    % Calculate control constants
    kp = (3 * w_cl^2 * T) / Lambda;
    ki = (w_cl^3 * T) / Lambda;
    kd = (3 * w_cl * T - 1) / Lambda;
end